packages/evm/src/memory.ts

```diff
@@ -63,14 +63,18 @@ export class Memory {
    * It fills up the difference between memory's length and `offset + size` with zeros.
    * @param offset - Starting position
    * @param size - How many bytes to read
+   * @param avoidCopy - Avoid memory copy if possible for performance reasons (optional)
    */
-  read(offset: number, size: number): Buffer {
+  read(offset: number, size: number, avoidCopy?: boolean): Buffer {
     this.extend(offset, size)
 
     const returnBuffer = Buffer.allocUnsafe(size)
     // Copy the stored "buffer" from memory into the return Buffer
 
     const loaded = this._store.slice(offset, offset + size)
+    if (avoidCopy === true && loaded.length === size) {
+      return loaded
+    }
```

Comment:
I think what we can do here is:

- Return the `slice` of the `_store`
- In case that `loaded.length` does not equal the size, expand the current internal `_store` with extra zeros. (Use the `extend` method to allocate in 8kb chunks)

Reply:
A that's a good idea, I will try that (have got 3 work hours from now on). I would nevertheless want to keep this in this limited scope (so "free" the memory one after one or - rather - over time eventually), otherwise this will get too hairy and uncontrollable.

In the current scope this feels already pretty good though, will also test on some couple of hundred tx loaded mainnet blocks.

Reply:
Direct update: ok, tested with `npm run client:start -- --discDns=false --discV4=false --maxPeers=0 --executeBlocks=1400000-1401000` on a couple of thousand txs, that all went well, only one fork (homestead) but already a good sign. 🙂

Reply:
Hmm. There is actually already a `this.extend(offset, size)` call at the beginning of the method.

I don't think it would have any effect if I just delete the zero filling conditional (not seeing how this can ever evaluate to true in the current setup of the method).

```typescript
    if (loaded.length < size) {
      // fill the remaining part of the Buffer with zeros
      returnBuffer.fill(0, loaded.length, size)
    }
```

Will do that.

Or maybe, from a procedural point of view: will *first* look into the current test failures, and *then* do it. 😋

Reply:
Ah right, sorry, there is indeed the `extend` method above.

Somewhat unrelated, but each time we `extend` and it creates a 8kb memory, we `concat` the Buffer which, I think (https://nodejs.org/api/buffer.html#static-method-bufferconcatlist-totallength) copies the previous Buffer as well, which we also might not want. I am not sure how other clients allocate these Buffers, you could split them up in chunks of whatever size without copying them, but then if you want to read over 2+ containers the read-logic gets a big complex (and you have to copy them again). (Think of memory in this "optimized" (?) case as a linked list?)

Reply:
This 96 is due to solidity, it initializes the "free memory pointer" at slot 0x40 (64) which will thus write 32 bytes at offset 64 and thus write 96. I am not sure if I am ok with directly initializing it to 96 bytes, since it is super likely that more memory writes are happening, and if solidity changes their initial memory structure we have this artifact laying around.